http://sourceware.org/ml/gdb-patches/2011-03/msg00946.html
Subject: [patch] ada: Fix rashes on gdbarch garbage

Hi Joel,

for gdb.ada/str_ref_cmp.exp on x86_64-fedora15-linux-gnu it turns:

(gdb) print String_Var (1 .. 3) = "Hel"
ERROR: Process no longer exists
UNRESOLVED: gdb.ada/str_ref_cmp.exp: print String_Var (1 .. 3) = "Hel"
->
(gdb) print String_Var (1 .. 3) = "Hel"
$1 = false
(gdb) FAIL: gdb.ada/str_ref_cmp.exp: print String_Var (1 .. 3) = "Hel"

GDB crashes the former way without this fix on:
	gcc-gnat-4.6.0-0.14.fc15.x86_64
	gcc-gnat-4.5.1-4.fc14.x86_64

I cannot test FSF gcc <=4.5 myself as its Ada build errors out for me on:
	uintp.adb:242:22: expect type "Hnum" defined at line 91

GDB PASSes on these GNATs even without this fix but that is not the point:
	GNAT 4.7.0 20110320 (experimental)
	GNAT 4.6.0 20110320 (prerelease)

This is only a suggestion, I do not know much about the ada-* code.


Thanks,
Jan


gdb/
2011-03-20  Jan Kratochvil  <jan.kratochvil@redhat.com>

	* ada-lang.c (replace_operator_with_call): Copy also GDBARCH.

--- a/gdb/ada-lang.c
+++ b/gdb/ada-lang.c
@@ -3661,6 +3661,7 @@ replace_operator_with_call (struct expression **expp, int pc, int nargs,
 
   newexp->nelts = exp->nelts + 7 - oplen;
   newexp->language_defn = exp->language_defn;
+  newexp->gdbarch = exp->gdbarch;
   memcpy (newexp->elts, exp->elts, EXP_ELEM_TO_BYTES (pc));
   memcpy (newexp->elts + pc + 7, exp->elts + pc + oplen,
           EXP_ELEM_TO_BYTES (exp->nelts - pc - oplen));

